pkg/apk/impl/const.go

```diff
@@ -0,0 +1,35 @@
+// Copyright 2022 Chainguard, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package impl
+
+const (
+	DefaultKeyRingPath       = "/etc/apk/keys"
+	DefaultSystemKeyRingPath = "/usr/share/apk/keys/"
+	indexFilename            = "APKINDEX.tar.gz"
+	// we are using these for fs.FS so should omit the leading /
+	reposFilePath     = "etc/apk/repositories"
+	archFilePath      = "etc/apk/arch"
+	keysDirPath       = "etc/apk/keys"
+	worldFilePath     = "etc/apk/world"
+	installedFilePath = "lib/apk/db/installed"
+	scriptsFilePath   = "lib/apk/db/scripts.tar"
+	scriptsTarPerms   = 0o644
+	triggersFilePath  = "lib/apk/db/triggers"
+
+	// for fetching the alpine keys
```

Comment:
I don't think we should depend on a random package for a set of keys, since they can rotate frequently.  Instead, we should talk with the Alpine project about implementing key bundles that can be fetched from a well-known location.

cc @ikke for ideas here.  Perhaps alpine-go can have a package for fetching keys?

Reply:
Yeah I wasn't enamoured of it either, but it got it to work. I would be very happy with a better alternative.

Reply:
Would that do more then just doing an HTTP GET from example here? https://alpinelinux.org/keys/

Reply:
@Ikke yeah I think that would work. 

The behaviour, though, is slightly different, in that `alpine-keys` package installs them all into `/usr/share/apk/keys/`, with symlinks for the relevant ones for each platform in `<arch>/`. Then the install places the ones just for the given platform in `/etc/apk/keys/`.

If I get them directly from https://alpinelinux.org/keys/ , I just have the keys, no platform-specific ones. Would I install them all in `/etc/apk/keys`?

Reply:
I'm thinking about proposing adding the keys that are used per arch to https://alpinelinux.org/releases.json, would that work?

Reply:
It would, I believe. Do you mean the actual keys, or just the names of the keys as references?

Reply:
names and urls where to find them.

Reply:
Sounds pretty good.

Should I wait for it? What's your estimate on when it will go in? And any way we can help?

Reply:
https://gitlab.alpinelinux.org/alpine/infra/alpine-mksite/-/merge_requests/58

Reply:
Cool, thanks. I signed up for notifications on that MR. Once that is in, we will consume it.

Reply:
I will also update gitlab.alpinelinux.org/alpine/go/releases to include this data

Reply:
@Ikke thanks for merging that in. I am not quite sure what the structure of `releases.json` is. I _think_ that what I need to do is, "look for value of `".keys"` in the `release_branches` array where `".rel_branch" == desired`, essentially, if I am looking for the aarch64 keys for 3.16:

```
curl https://alpinelinux.org/releases.json | jq -r '.release_branches[] | select(.rel_branch == "v3.16") | .keys["aarch64"][].url'
```

Is that correct?

Reply:
OK, got it to work, all looks good. No longer trying to install arbitrary `alpine-keys` package.

I will mark this as "Resolved", but @Ikke you might want to look at the implementation in `pkg/apk/impl/`; we can upstream some stuff from there.

Reply:
@Ikke is there someone I can find a reference on signature validation? It looks like the algorithm changed somewhere down the line. I see the larger vs smaller keys (451 bytes in the PEM file vs 800 bytes, which I assume are the 4096 RSA keys), and larger vs smaller signatures (512 vs 256). Yet was the algorithm different prior to the longer keys? We are using a fairly straightforward logic that works for the 4096, but not for the older keys. Simplified below:

```go
func RSAVerifySHA1Digest(sha1Digest, signature []byte, publicKey []byte) error {
	block, _ := pem.Decode(publicKey)
	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	rsaPub, ok := pub.(*rsa.PublicKey)
	return rsa.VerifyPKCS1v15(rsaPub, crypto.SHA1, sha1Digest, signature)
}
```

Reply:
FWIW file sizes. I can verify the `aarch64.pub` and `aarch64.sig` but not the `riscv64` counterparts.

```
-rw-r--r--   1 user  wheel     800 Jan  4 12:47 aarch64.pub
-rw-r--r--   1 user  wheel     512 Jan  4 13:01 aarch64.sig
-rw-r--r--   1 user  wheel     451 Jan  4 12:47 riscv.pub
-rw-r--r--   1 user  wheel     256 Jan  4 13:01 riscv64.sig
```

Reply:
I'm not aware of any algorithm change, only that we recently created new 4096-bits keys. As riscv64 was generated on a different moment. `abuild-keygen` uses `openssl genrsa`, so we'd have to look at openssl to see if anything changed.

Reply:
OK. If there are pointers you can share, it would be appreciated.

Reply:
Is this what you mean?

riscv64:
```
$ openssl rsa -pubin -in alpine-devel@lists.alpinelinux.org-60ac2099.rsa.pub -noout -text | head -n1
Public-Key: (2048 bit)
```

aarch64:
```
$ openssl rsa -pubin -in alpine-devel@lists.alpinelinux.org-616ae350.rsa.pub -noout -text | head -n1                                                                     
Public-Key: (4096 bit)
```

riscv64 is a 2048-bits key, while aarch64 is a 4096-bits key.